src/tasks/LibraryBuilder/Templates/autoinit.c

```diff
@@ -0,0 +1,162 @@
+
+    int len = strlen (name);
+    int has_extension = len > 3 && name [len - 4] == '.' && (!strcmp ("exe", name + (len - 3)) || !strcmp ("dll", name + (len - 3)));
+
+    // add extensions if required.
+    strlcpy (filename, name, sizeof (filename));
+    if (!has_extension) {
+        strlcat (filename, ".dll", sizeof (filename));
+    }
+
+    if (culture && strcmp (culture, ""))
+        res = snprintf (path, sizeof (path) - 1, "%s/%s/%s", bundle_path, culture, filename);
+    else
+        res = snprintf (path, sizeof (path) - 1, "%s/%s", bundle_path, filename);
+    assert (res > 0);
+
+    struct stat buffer;
+    if (stat (path, &buffer) == 0) {
+        MonoAssembly *assembly = mono_assembly_open (path, NULL);
+        assert (assembly);
+        return assembly;
+    }
+    return NULL;
+}
+
+static MonoAssembly*
+mono_assembly_preload_hook (MonoAssemblyName *aname, char **assemblies_path, void* user_data)
+{
+    const char *name = mono_assembly_name_get_name (aname);
+    const char *culture = mono_assembly_name_get_culture (aname);
+    return mono_load_assembly (name, culture);
+}
+
+static unsigned char *
+load_aot_data (MonoAssembly *assembly, int size, void *user_data, void **out_handle)
+{
+    *out_handle = NULL;
+
+    char path [1024];
+    int res;
+
+    MonoAssemblyName *assembly_name = mono_assembly_get_name (assembly);
+    const char *aname = mono_assembly_name_get_name (assembly_name);
+
+    res = snprintf (path, sizeof (path) - 1, "%s/%s.aotdata", bundle_path, aname);
+    assert (res > 0);
+
+    int fd = open (path, O_RDONLY);
+    if (fd < 0) {
+        return NULL;
+    }
+
+    void *ptr = mmap (NULL, size, PROT_READ, MAP_FILE | MAP_PRIVATE, fd, 0);
+    if (ptr == MAP_FAILED) {
+        close (fd);
+        return NULL;
+    }
+
+    close (fd);
+    *out_handle = ptr;
+    return (unsigned char *) ptr;
+}
+
+static void
+free_aot_data (MonoAssembly *assembly, int size, void *user_data, void *handle)
+{
+    munmap (handle, size);
+}
+
+void
+runtime_init_callback ()
+{
+    initialize_runtimeconfig ();
+
+    initialize_appctx_env_variables ();
+
+    register_aot_modules ();
+
+    // register all bundled modules
+
+    mono_set_assemblies_path ((bundle_path && bundle_path[0] != '\0') ? bundle_path : "./");
+
+    mono_jit_set_aot_only (true);
+
+    mono_install_assembly_preload_hook (mono_assembly_preload_hook, NULL);
+
+    // TODO test debug scenario
+#if DEBUG_ENABLED
+    bool wait_for_debugger = false;
+    mono_debug_init (MONO_DEBUG_FORMAT_MONO);
+    if (wait_for_debugger) {
+        char* options[] = { "--debugger-agent=transport=dt_socket,server=y,address=0.0.0.0:55555" };
+        mono_jit_parse_options (1, options);
+    }
+#endif
+
+    mono_install_load_aot_data_hook (load_aot_data, free_aot_data, NULL);
+
+    mono_set_signal_chaining (true);
+
+    mono_jit_init ("mono.self.contained.library"); // Pass in via LibraryBuilder?
+
+%ASSEMBLIES_LOADER%
+}
+
+void
+init_mono_runtime ()
```

Comment:
I think this method should have a different name, you get the impression that its initialize the runtime, but it just sets up the callback hook, you should just do that in the autoinit constructor function directly.

Reply:
Removed this intermediate wrapper